In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

In [2]:
train = pd.read_csv('../data/train-parking.csv', parse_dates=[[3,4]], infer_datetime_format=True)
test=pd.read_csv('../data/test-no-label-parking.csv', parse_dates=[[3,4]], infer_datetime_format=True)
valid_dow = pd.read_csv('../data/valid_dow.csv', infer_datetime_format=True, index_col='index').set_index('index.1')
valid_hg = pd.read_csv('../data/valid_hg.csv', infer_datetime_format=True, index_col='index').set_index('index.1')
valid_swhg = pd.read_csv('../data/valid_swhg.csv', infer_datetime_format=True, index_col='index').set_index('index.1')

In [3]:
train.head()

Date_Time           Street                  From             To  \
0 2014-01-07 16:19:00   Mission Street           25th Street    26th Street   
1 2014-01-18 20:42:00      Polk Street          Ellis Street   Olive Street   
2 2014-01-18 20:39:00  Van Ness Avenue       Geary Boulevard  Myrtle Street   
3 2014-01-18 20:38:00  Van Ness Avenue           Bush Street    Fern Street   
4 2014-01-18 20:38:00  Van Ness Avenue  Daniel Burnham Court    Post Street   

   Real.Spots  Street.Length  any_spot  
0           4     179.132970         1  
1           0      52.740210         0  
2           0      52.517840         0  
3           0      52.405315         0  
4           0      52.191193         0

##### Get the proper train and validation set

In [4]:
train_dow = train[~train.index.isin(valid_dow.index)]
train_hg = train[~train.index.isin(valid_hg.index)]
train_swhg = train[~train.index.isin(valid_swhg.index)]

## Select validation set - round 1

In [5]:
# get long and lat for train street
st_latlng = pd.read_csv("../data/train_longlat.csv", index_col = False,  dtype = {'lat':np.float64,'lng':np.float64})

In [6]:
train_hg = pd.merge(train_hg,st_latlng,how ='left',left_on=["Street","From","To"], right_on=["Street","From","To"])
valid_hg= pd.merge(valid_hg,st_latlng,how ='left',left_on=["Street","From","To"], right_on=["Street","From","To"])

In [7]:
# sensor related data
sensor = pd.read_csv('../data/aggregated_sensor.csv')

In [8]:
sensor['TIME_OF_DAY'] = sensor['TIME_OF_DAY']/100
sensor['TIME_OF_DAY'] = sensor['TIME_OF_DAY'].astype('int')

In [9]:
# Clustering
from sklearn.cluster import KMeans
spots = np.vstack(train_hg[['lat','lng']].values)
kmeans = KMeans().fit(spots)
train_hg['street_cluster'] = kmeans.predict(train_hg[['lat', 'lng']])
valid_hg['street_cluster'] = kmeans.predict(valid_hg[['lat', 'lng']])

In [10]:
# Time features
train_hg['Minute'] = train_hg.Date_Time.dt.minute
valid_hg['Minute'] = pd.to_datetime(valid_hg['Date_Time']).dt.minute
train_hg['Hour'] = train_hg.Date_Time.dt.hour
valid_hg['Hour'] = pd.to_datetime(valid_hg['Date_Time']).dt.hour
train_hg['Dow'] = train_hg.Date_Time.dt.weekday
valid_hg['Dow'] = pd.to_datetime(valid_hg['Date_Time']).dt.weekday

In [11]:
# Convert the time (Hour, Minute) to a numerical value XXXX
train_hg['Time'] = train_hg["Hour"]*100 + train_hg['Minute']
valid_hg['Time'] = valid_hg["Hour"]*100 + valid_hg['Minute']

In [12]:
def day_type(x):
    if x == 6 or x == 0:
        return 1
    else:
        return 0
    
train_hg['isweekend'] = train_hg['Dow'].apply(day_type)
valid_hg['isweekend'] = valid_hg['Dow'].apply(day_type)

In [13]:
# Single mean encoding
from sklearn.model_selection import KFold

def multi_reg_mean_encoding(train, cols, splits =5):
    kf = KFold(n_splits = splits)
    global_mean = train.any_spot.mean()
    
    for col in cols:
        means = train.groupby(col).any_spot.mean()
        train[col+'_mean_enc'] = train[col].map(means)        
        for tr_ind,val_ind in kf.split(train):
            tr,val = train.iloc[tr_ind],train.iloc[val_ind]
            foldmeans = tr.groupby(col).any_spot.mean()
            train.loc[val_ind,col+"_mean_enc"] = train.loc[val_ind,col].map(foldmeans)    
        train[col+"_mean_enc"].fillna(global_mean,inplace=True)   
        
#mean encoding for validation and test data
def multi_test_mean_encoding(test, train, cols):
    for col in cols:
        global_mean = train.any_spot.mean()
        means = train.groupby(col).any_spot.mean()
        test[col+"_mean_enc"] = test[col].map(means)
        test[col+"_mean_enc"].fillna(global_mean, inplace=True)

In [14]:
#create a column for From_To pair
train_hg['From_To'] = train_hg['From'] + '_' + train_hg['To']
valid_hg['From_To'] = valid_hg['From'] + '_' + valid_hg['To']

In [15]:
cols = ['Dow','isweekend','From_To','Street','Hour', 'street_cluster' ]
multi_reg_mean_encoding(train_hg, cols, splits =5)
multi_test_mean_encoding(valid_hg, train_hg, cols)

In [16]:
# Double variable mean encoding
colpairs = [('street_cluster','Dow'),('Dow','Hour'), ('street_cluster','Hour'),
            ('street_cluster','isweekend'),('From_To','Dow'),('From_To','Hour')]
pairlist = []
for pair in colpairs:
    colname = pair[0] + '_' + pair[1]
    pairlist.append(colname)
    train_hg[colname] = list(zip(train_hg[pair[0]], train_hg[pair[1]]))
    valid_hg[colname] = list(zip(valid_hg[pair[0]], valid_hg[pair[1]]))

In [17]:
multi_reg_mean_encoding(train_hg, pairlist, splits =5)
multi_test_mean_encoding(valid_hg, train_hg, pairlist)

In [18]:
#drop Data_Time, Real.Spots, street_cluster_Dow, Dow_Hour, street_cluster_Hour, street_cluster_isweekend, From_To_Dow, From_To_Hour
train_hg = train_hg.drop(['Date_Time','Real.Spots', 'street_cluster_Dow', 'Dow_Hour', 'street_cluster_Hour', 'street_cluster_isweekend', 'From_To_Dow', 'From_To_Hour'], axis =1)
valid_hg = valid_hg.drop(['Date_Time', 'Real.Spots', 'street_cluster_Dow', 'Dow_Hour', 'street_cluster_Hour', 'street_cluster_isweekend', 'From_To_Dow', 'From_To_Hour'], axis = 1)

In [19]:
train_hg['Street'] = train_hg.Street.str.lower()
train_hg['From'] = train_hg.From.str.lower()
train_hg['To'] = train_hg.To.str.lower()
valid_hg['Street'] = valid_hg.Street.str.lower()
valid_hg['From'] = valid_hg.From.str.lower()
valid_hg['To'] = valid_hg.To.str.lower()

In [20]:
# Join sensor data:
train_hg = pd.merge(train_hg,sensor, how='left',left_on=['Street','Dow','Hour'], right_on=['STREET_NAME','DOW','TIME_OF_DAY'])
valid_hg = pd.merge(valid_hg,sensor, how='left',left_on=['Street','Dow','Hour'], right_on=['STREET_NAME','DOW','TIME_OF_DAY'])

train_hg = train_hg.drop(['STREET_NAME','DOW','TIME_OF_DAY'],axis =1)
valid_hg = valid_hg.drop(['STREET_NAME','DOW','TIME_OF_DAY'],axis =1)

In [21]:
# convert categorical variables to code
def process_dfs(train_df, test_df, cols):
    # transfer categorical 
    for name in cols:
        train_df[name]=train_df[name].astype('category').cat.as_ordered()
        test_df[name] = pd.Categorical(test_df[name], categories=train_df[name].cat.categories, ordered=True)
    return train_df, test_df
  
def cat2code(train_df, test_df, cols):
    # transfer categorical 
    for name in cols:
        train_df[name]= train_df[name].cat.codes
        test_df[name] = test_df[name].cat.codes
    return train_df, test_df

In [22]:
catcols = ['Street','From','To', 'From_To']
train_hg,valid_hg = process_dfs(train_hg.copy(),valid_hg.copy(),catcols)
train_hg,valid_hg = cat2code(train_hg,valid_hg,catcols)

In [23]:
# split X and y
def split_response(df, response):
    y = df[response].values
    df.drop([response],axis=1,inplace=True)
    return df,y

In [24]:
X_train_hg,y_train_hg = split_response(train_hg.copy(),'any_spot')
X_valid_hg,y_valid_hg = split_response(valid_hg.copy(),'any_spot')

In [25]:
def f05_score_soft(labels, preds):
    tp = np.sum((labels==(preds>0.5)) & (labels==1))
    tn = np.sum((labels==(preds<0.5)) & (labels==0))
    fp = np.sum((preds>0.5))-tp
    fn = np.sum(preds<0.5)-tn
    p = tp/(tp+fp)
    r = tp/(tp+fn)
    score = 1.25*p*r/(0.25*p+r)
    return score

def f05_score_hard(labels, preds):
    tp = np.sum((labels==preds) & (labels==1))
    tn = np.sum((labels==preds) & (labels==0))
    fp = np.sum(preds==1)-tp
    fn = np.sum(preds==0)-tn
#    print(tp, tn, fp, fn)
    p = tp*1.0/(tp+fp)
    r = tp*1.0/(tp+fn)
    score = 1.25*p*r/(0.25*p+r)
    return score

In [26]:
# Manual gridsearch using validation set
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import precision_score, recall_score, accuracy_score
import itertools

n_estimators = [20, 40, 60, 80]
max_depth = [3,5,10,15,20,25,30]
max_features = [0.2,0.4,0.5,0.6,0.8,0.9,1]
weight = [0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.3, 1.4, 1.5]
para_lists = [n_estimators, max_depth, max_features, weight]
combo = list(itertools.product(*para_lists))

In [27]:
parameters = []
f05 = []
precision = []
recall = []

for p in combo:
    np.random.seed(42)
    m = RandomForestClassifier(n_estimators = p[0], max_depth =p[1], max_features=p[2],class_weight={0:1,1:p[3]}, min_samples_split = 2, n_jobs=-1)
    m.fit(X_train_hg,y_train_hg)
    pred = m.predict(X_valid_hg)
    prec = precision_score(y_valid_hg, pred)
    rec = recall_score(y_valid_hg, pred)
    f = f05_score_hard(y_valid_hg, pred)
    parameters.append(p)
    f05.append(f)
    precision.append(prec)
    recall.append(rec)

table2 = pd.DataFrame({'Parameters': parameters, 'f0.5': f05, 'precision':precision, 'recall':recall})

In [28]:
table2.sort_values('f0.5', ascending=False, inplace=True)

In [29]:
table2.head()

Parameters      f0.5  precision    recall
818    (40, 25, 1, 1.5)  0.583039   0.647059  0.417722
881    (40, 30, 1, 1.5)  0.583039   0.647059  0.417722
522   (40, 5, 0.5, 0.7)  0.581395   0.735294  0.316456
688    (40, 15, 1, 1.1)  0.573477   0.640000  0.405063
263  (20, 20, 0.4, 0.9)  0.572755   0.606557  0.468354

In [30]:
# use the best hyperparameters (40, 25, 1, 1.5) to get feature importance:
np.random.seed(42)
m2 = RandomForestClassifier(n_estimators = 40, max_depth =25, max_features=1, class_weight={0:1,1:1.5}, min_samples_split = 2, n_jobs=-1)
m2.fit(X_train_hg,y_train_hg)

feat_imp_hg = pd.DataFrame({
    'features': X_train_hg.columns,
    'imp' :m2.feature_importances_
})
feat_imp_hg.sort_values(by='imp', inplace=True, ascending=False)

In [31]:
feat_imp_hg

features       imp
3                       Street.Length  0.049633
10                               Time  0.045867
7                              Minute  0.044440
15                   From_To_mean_enc  0.041422
12                            From_To  0.038419
4                                 lat  0.034857
5                                 lng  0.034227
1                                From  0.031935
2                                  To  0.030489
25            TOTAL_OCCUPIED_TIME_pct  0.025186
16                    Street_mean_enc  0.025136
20                  Dow_Hour_mean_enc  0.025048
23               From_To_Dow_mean_enc  0.024776
26              TOTAL_VACANT_TIME_pct  0.024359
21       street_cluster_Hour_mean_enc  0.024331
31            NONOP_OCCUPIED_TIME_pct  0.023317
19        street_cluster_Dow_mean_enc  0.021856
27             TOTAL_UNKNOWN_TIME_pct  0.021167
6                      street_cluster  0.021158
22  street_cluster_isweekend_mean_enc  0.020796
8                                Hour  0.020535
24              From_To_Hour_mean_enc  0.020423
32              NONOP_VACANT_TIME_pct  0.019674
18            street_cluster_mean_enc  0.019516
34              GMP_OCCUPIED_TIME_pct  0.019502
33             NONOP_UNKNOWN_TIME_pct  0.019206
17                      Hour_mean_enc  0.018795
36               GMP_UNKNOWN_TIME_pct  0.018751
13                       Dow_mean_enc  0.017981
40                   TOTAL_TIME_spots  0.017878
28               OP_OCCUPIED_TIME_pct  0.017028
42                   NONOP_TIME_spots  0.015828
29                 OP_VACANT_TIME_pct  0.015415
35                GMP_VACANT_TIME_pct  0.015341
30                OP_UNKNOWN_TIME_pct  0.015326
43                     GMP_TIME_spots  0.014622
41                      OP_TIME_spots  0.014058
14                 isweekend_mean_enc  0.013934
38               COMM_VACANT_TIME_pct  0.013521
9                                 Dow  0.013521
0                              Street  0.013519
37             COMM_OCCUPIED_TIME_pct  0.013038
39              COMM_UNKNOWN_TIME_pct  0.012272
44                    COMM_TIME_spots  0.006272
11                          isweekend  0.005624

In [32]:
# drop feature if FI<0.015
to_drop_hg = feat_imp_hg[feat_imp_hg['imp'] <= 0.015].features

### DOW-retrain entire model - remember to drop insignificant columns in the end

In [33]:
train = pd.merge(train,st_latlng,how ='left',left_on=["Street","From","To"], right_on=["Street","From","To"])
test= pd.merge(test,st_latlng,how ='left',left_on=["Street","From","To"], right_on=["Street","From","To"])

spots = np.vstack(train[['lat','lng']].values)
kmeans = KMeans().fit(spots)
train['street_cluster'] = kmeans.predict(train[['lat', 'lng']])
test['street_cluster'] = kmeans.predict(test[['lat', 'lng']])

In [34]:
train['Minute'] = train.Date_Time.dt.minute
test['Minute'] = pd.to_datetime(test['Date_Time']).dt.minute
train['Hour'] = train.Date_Time.dt.hour
test['Hour'] = pd.to_datetime(test['Date_Time']).dt.hour
train['Dow'] = train.Date_Time.dt.weekday
test['Dow'] = pd.to_datetime(test['Date_Time']).dt.weekday

In [35]:
train['Time'] = train["Hour"]*100 + train['Minute']
test['Time'] = test["Hour"]*100 + test['Minute']
train['isweekend'] = train['Dow'].apply(day_type)
test['isweekend'] = test['Dow'].apply(day_type)
train['From_To'] = train['From'] + '_' + train['To']
test['From_To'] = test['From'] + '_' + test['To']

In [36]:
cols = ['Dow','isweekend','From_To','Street','Hour', 'street_cluster' ]
multi_reg_mean_encoding(train, cols, splits =5)
multi_test_mean_encoding(test, train, cols)

In [37]:
colpairs = [('street_cluster','Dow'),('Dow','Hour'), ('street_cluster','Hour'),
            ('street_cluster','isweekend'),('From_To','Dow'),('From_To','Hour')]
pairlist = []
for pair in colpairs:
    colname = pair[0] + '_' + pair[1]
    pairlist.append(colname)
    train[colname] = list(zip(train[pair[0]], train[pair[1]]))
    test[colname] = list(zip(test[pair[0]], test[pair[1]]))

In [38]:
multi_reg_mean_encoding(train, pairlist, splits =5)
multi_test_mean_encoding(test, train, pairlist)

In [39]:
train = train.drop(['Date_Time','Real.Spots', 'street_cluster_Dow', 'Dow_Hour', 'street_cluster_Hour', 'street_cluster_isweekend', 'From_To_Dow', 'From_To_Hour'], axis =1)
test = test.drop(['Date_Time', 'street_cluster_Dow', 'Dow_Hour', 'street_cluster_Hour', 'street_cluster_isweekend', 'From_To_Dow', 'From_To_Hour'], axis = 1)

In [40]:
train['Street'] = train.Street.str.lower()
train['From'] = train.From.str.lower()
train['To'] = train.To.str.lower()
test['Street'] = test.Street.str.lower()
test['From'] = test.From.str.lower()
test['To'] = test.To.str.lower()

In [42]:
train = pd.merge(train,sensor, how='left',left_on=['Street','Dow','Hour'], right_on=['STREET_NAME','DOW','TIME_OF_DAY'])
test = pd.merge(test,sensor, how='left',left_on=['Street','Dow','Hour'], right_on=['STREET_NAME','DOW','TIME_OF_DAY'])

train = train.drop(['STREET_NAME','DOW','TIME_OF_DAY'],axis =1)
test = test.drop(['STREET_NAME','DOW','TIME_OF_DAY'],axis =1)

In [43]:
catcols = ['Street','From','To', 'From_To']
train,test = process_dfs(train.copy(),test.copy(),catcols)
train,test = cat2code(train,test,catcols)

In [44]:
X_train,y_train = split_response(train.copy(),'any_spot')

In [45]:
# drop unimportant features
X_train.drop(to_drop_hg, axis = 1, inplace=True)
test.drop(to_drop_hg, axis = 1, inplace=True)

In [47]:
# Fit rf
np.random.seed(42)
m_hg = RandomForestClassifier(n_estimators = 40, max_depth =25, max_features=1, class_weight={0:1,1:1.5}, min_samples_split = 2, n_jobs=-1)
m_hg.fit(X_train,y_train)
pred_hg = m_hg.predict(test)

In [48]:
result = pd.DataFrame({"id": np.arange(1,len(test)+1),"any_spot":pred_hg})

In [49]:
result[['id','any_spot']].to_csv("../submissions/32_hg_test.csv", index =False)

By Hour Group: 

Result from leaderboard: 0.51987

on validation set: 0.583039